docs/developer/architecture/centralized-cli.rst

```diff
@@ -0,0 +1,145 @@
+- Solution Administration (CRUD Environment, Import/Remove Solution, ...)
+- Cluster Service Administration (Configure Dex, Prometheus, Alert Manager,
+  ...)
+
+K8s provides the kubectl CLI, which is distributed and enables all kind of
+interaction with all Kubernetes resources, through apiserver, but its usage
+often requires to build verbose YAML files.
+
+Currently, MetalK8s provides other set of scripts or manual procedures, but
+those are located in various locations, their usage may vary and they are not
+developed using the same logic.
+
+This makes the CLI and associated documentation not super intuitive and it
+also makes the maintenance more expensive in the long term.
+
+The goal of the project is to provide MetalK8s administrator with an intuitive
+and easy to use set of tools in order to administrate and operate a finite set
+of functionalities.
+
+Because kubectl CLI is already in place and is well known by Kubernetes
+administrators, it will be used as a standard to follow for all other MetalK8s
+CLIs:
+
+- CLI follows kubectl style: kubectl <action> <resource>
+- CLI provides an exhaustive help, per action, with relevant examples
+- CLI is not interactive (except for static user provisioning)
+- CLI should not require password input
+- CLI provides a dryrun mode for intrusive operations
+- CLI provides a verbose (or debug) mode
+- CLI implementation relies on secure APIs
+- CLI support <action> completion for easy discovery
+- CLI output is standardized and human readable by default
+- CLI output can be formatted in JSON or YAML
+- CLI can be executed from outside of the cluster (except maybe for very first
+  install script)
+
+When it is possible, it would make sense to leverage kubectl plugin
+
+Requirements
+------------
+
+Cluster Resources Administration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+tools: kubectl
+
++------------+------------------+-----------------------------------------+
+| Resource   | action           | parameters                              |
++============+==================+=========================================+
+| node       | add              | name, ssh-user, hostname or ip,         |
+|            |                  | ssh port, ssh-key-path, sudo-required,  |
+|            |                  | roles, manifest-template-path           |
++------------+------------------+-----------------------------------------+
+| the manifest-template-path is the path to the manifest describing the   |
+| resource. Everything that is in the command line, override the manifest |
+| content. This apply to Node and Volumes.                                |
++------------+------------------+-----------------------------------------+
+| node       | edit             | name, hostname (to change node hostname)|
++------------+------------------+-----------------------------------------+
+| node       | deploy(or apply?)| <list of nodes>, dry-run                |
```

Comment:
Hmm it's not part of `kubectl`

Reply:
isn't it the kind of thing we can achieve with the kubectl plugin? i.e. enrich kubectl with new functionalities? but from a usage point of view, the administrator will invoke kubectl?

Reply:
Hmm for this one is exacly one we do NOT want to put in a plugin because it's an interaction with salt directly so not part of kubectl at all

Reply:
(can't review the whole PR atm, but just my 2 cents on this question)

We **can** call SaltAPI from `kubectl`. We **could** also send commands in the `salt-master` container. Nothing prevents us from interacting with Salt from a `kubectl` plugin.

There are some commands which will require direct interaction on the host however, maybe because the operation will trigger a master restart (e.g. upgrade/downgrade), or because it does not make sense to do it from somewhere else (e.g. managing Metal or Solution ISOs on the Bootstrap). But deploying a Node is a standard operation you may want to do from outside the cluster (we do it from the UI already).

Reply:
Of course we can do it from a kubectl plugin but for me it's a bad idea. Kubectl plugin shouldn't interact with salt IMO.

> There are some commands which will require direct interaction on the host however, maybe because the operation will trigger a master restart (e.g. upgrade/downgrade), or because it does not make sense to do it from somewhere else (e.g. managing Metal or Solution ISOs on the Bootstrap). But deploying a Node is a standard operation you may want to do from outside the cluster (we do it from the UI already).

???
I do not see why it's related to kubectl plugin or not ?? You can execute your kubectl plugin from outside the cluster, you can also execute our own CLI from outside the cluster

Reply:
> you can also execute our own CLI from outside the cluster

No. This "extra CLI" we'll build for Bootstrap-local operations will not be a "redistributable package", just an executable in an ISO.

Otherwise, how do you allow this CLI to authenticate against Salt without a kubeconfig? Would you reimplement the login dance with K8s API to then be able to reach Salt API?
Also, what's the benefit of having two CLIs if they can both work anywhere?

Reply:
> No. This "extra CLI" we'll build for Bootstrap-local operations will not be a "redistributable package", just an executable in an ISO.

So for you al extral CLI are for stuff that cannot go outside the cluster otherwise it's a kubectl plugin ??
For me no.

To add a node you can perfectly have a CLI that work outside of the cluster or a kubectl plugin but IMO kubectl plugins shouldn't interact with salt

Reply:
@TeddyAndrieux That's the second time you say `IMO kubectl plugins shouldn't interact with salt`, do you have reason for that?

I've no opinion on that part, but giving your argument for this could help convince people (or at least understand your point of view).

Reply:
Why not, @TeddyAndrieux ? There's SaltAPI they can work with, using the `KubeConfig` token.

Now, to be clear: `kubectl` plugins *only allow to add top-level commands*. So you *can't* write a plugin which all of a sudden would give you a `kubectl node deploy` command or something along those lines. You could have `kubectl metalk8s deploy node`.

Reply:
OK, in the above: you *could* have `kubectl node deploy` because `kubectl node` doesn't exist., but that's mere coincidence and I think it'd be very bad style (and potential future-incompatibility) to hijack random nouns.

Reply:
For me kubectl plugins should only act on apiserver and not on something we deploy on our side, for me it's something that can work on every kubernetes deployment ...

I agree and I know everything is doable from kubectl plugin but for me it's not a good idea. Anyway as you want

Reply:
okay so If I understand correctly, with kubectl, we can add a new action but we can't override an existing one (like kubectl add node, but with additional arguments)?

Reply:
> for me it's something that can work on every kubernetes deployment 

Not really. Core `kubectl` functionalities, yes. But **plugins** are "additional functionality". For instance, we could have a `kubectl-prom` plugin which helps manage whatever comes with the Prometheus Operator.

> with kubectl, we can add a new action but we can't override an existing one (like kubectl add node, but with additional arguments)?

The way plugins work is quite simple. Put a binary called `kubectl-<plugin-name>` in the right place, and whenever you will run `kubectl <plugin-name> ...`, the call will be forwarded to this binary by `kubectl`, nothing more.
So indeed, there's no way to override existing `kubectl` functionality with plugins, only to extend it. In our case, as Nicolas said, we'll likely have all our commands under `kubectl metalk8s ...`.

Reply:
Okay, thanks for the explanation ... I think I understand now